cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/DataPipelineApp.java

```diff
@@ -57,6 +57,12 @@ public void configure() {
     ETLBatchConfig config = getConfig();
     setDescription(Objects.firstNonNull(config.getDescription(), DEFAULT_DESCRIPTION));
 
+    // Note: the generateSpec(config) will register the plugin usage at application configurer level. Later during
+    // the deployment phase when the individual programs are being added DefaultAppConfigurer will also call
+    // addDatasetsAndPlugins on the plugins registered in the individual program to register them at
+    // ApplicationConfigurer level. This is needed because when a plugin is used in custom cdap program, the plugin is
```

Comment:
I don't understand. Why plugins also need to be added to the app spec? If plugin are already stored in the program spec, isn't the program runtime service should inspect it and find the artifacts to localize?

Reply:
Storing all the plugin information at the app level can be avoided and also with current changes it will be possible to localize only the needed plugin. Although to keep the changes minimum it has not be done in this PR. I have opened a jira to track this here: https://issues.cask.co/browse/CDAP-14188

Reply:
This comment isn't something the app should know and should be removed. Since ApplicationSpecification has a getPlugins() method, we still need to adhere to the contract there. Every plugin in a program should be added to the AppSpec (I thought that's what the rest of this PR is doing?).

Generating the spec, which registers the plugins, should be done in the workflow so that all the plugins are associated with the workflow.

It is good to have a comment, but it should be something more like:

Plugin usage is registered in the workflow so that CDAP knows what plugins the workflow needs to run. If a plugin has a requirement that will not be available during that run, CDAP can fail the run early, before provisioning is performed. If plugins were registered in the application, CDAP would not be able to fail the run early.

Reply:
> This comment isn't something the app should know and should be removed. Since ApplicationSpecification has a getPlugins() method, we still need to adhere to the contract there. Every plugin in a program should be added to the AppSpec (I thought that's what the rest of this PR is doing?).

Removed the comments. I missed that we expose all plugins from the ApplicationSpecification through public api. If we decide not to store it then we will need to deprecate it. I will make a note on the link jira above. The PR changes register the plugins at for DataPipelineApp at program level and with the existing code which collects plugin information from program this bubbles up to app level. This was not happening before for the DataPipelineApp as it did not explicitly call usePlugins from programs although the end to end functionality was working fine since BatchPipelineSpecGenerator was registering all the plugins to app level internally. That has been moved to Workflow now and the actual plugin information which is stored at app level consists of plugin information collected from programs.

> Generating the spec, which registers the plugins, should be done in the workflow so that all the plugins are associated with the workflow.

Done.

> It is good to have a comment, but it should be something more like:
Plugin usage is registered in the workflow so that CDAP knows what plugins the workflow needs to run. If a plugin has a requirement that will not be available during that run, CDAP can fail the run early, before provisioning is performed. If plugins were registered in the application, CDAP would not be able to fail the run early.

Done